solr/solrj/src/java/org/apache/solr/client/solrj/impl/HttpSolrJdkClient.java

```diff
@@ -0,0 +1,449 @@
+import org.apache.solr.common.util.NamedList;
+import org.apache.solr.common.util.ObjectReleaseTracker;
+import org.apache.solr.common.util.SolrNamedThreadFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class HttpSolrJdkClient extends HttpSolrClientBase {
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+  private static final String USER_AGENT =
+      "Solr[" + MethodHandles.lookup().lookupClass().getName() + "] 1.0";
+
+  protected HttpClient httpClient;
+
+  protected ExecutorService executor;
+
+  private boolean shutdownExecutor;
+
+  protected HttpSolrJdkClient(String serverBaseUrl, HttpSolrJdkClient.Builder builder) {
+    super(serverBaseUrl, builder);
+
+    HttpClient.Redirect followRedirects =
+        Boolean.TRUE.equals(builder.followRedirects)
+            ? HttpClient.Redirect.NORMAL
+            : HttpClient.Redirect.NEVER;
+    HttpClient.Builder b = HttpClient.newBuilder().followRedirects(followRedirects);
+    if (builder.sslContext != null) {
+      b.sslContext(builder.sslContext);
+    }
+
+    if (builder.executor != null) {
+      this.executor = builder.executor;
+      this.shutdownExecutor = false;
+    } else {
+      BlockingQueue<Runnable> queue = new LinkedBlockingQueue<>(1024);
+      this.executor =
+          new ExecutorUtil.MDCAwareThreadPoolExecutor(
+              4,
+              256,
+              60,
+              TimeUnit.SECONDS,
+              queue,
+              new SolrNamedThreadFactory(this.getClass().getSimpleName()));
+      this.shutdownExecutor = true;
+    }
+    b.executor(this.executor);
+
+    if (builder.useHttp1_1) {
+      b.version(HttpClient.Version.HTTP_1_1);
+    }
+
+    if (builder.cookieHandler != null) {
+      b.cookieHandler(builder.cookieHandler);
+    }
+
+    if (builder.proxyHost != null) {
+      if (builder.proxyIsSocks4) {
+        log.warn(
+            "Socks4 is likely not supported by this client.  See https://bugs.openjdk.org/browse/JDK-8214516");
+      }
+      b.proxy(ProxySelector.of(new InetSocketAddress(builder.proxyHost, builder.proxyPort)));
+    }
+    this.httpClient = b.build();
+    updateDefaultMimeTypeForParser();
+
+    // This is a workaround for the case where the first request using
+    // this client:
+    //
+    // (1) no SSL/TLS (2) using POST with stream and (3) using Http/2
+    //
+    // The JDK Http Client will send an upgrade request over Http/1
+    // along with request content in the same request.  However,
+    // the Jetty Server underpinning Solr does not accept this.
+    //
+    // By sending a ping before any real requests occur, the client
+    // knows if Solr can accept Http/2, and no additional
+    // upgrade requests will be sent.
+    //
+    // See https://bugs.openjdk.org/browse/JDK-8287589
+    // See https://github.com/jetty/jetty.project/issues/9998#issuecomment-1614216870
+    //
+    try {
+      ping();
```

Comment:
> Second, I am calling "ping" in the constructor, which I felt was a lesser evil than having to check a volatile boolean with each request to see if "ping" was called yet.

It would be somewhat unexpected to have the client start sending calls to the server by just constructing it. May it even cause real issues since startup order between a customer client application and the solr server would then become significant?